src/unit_tests/remoted/test_sendmsg.c

```diff
@@ -39,32 +39,166 @@ static int group_teardown(void ** state) {
     test_mode = 1;
 
     keyentry *node_key = NULL;
     os_calloc(1, sizeof(keyentry), node_key);
-    node_key->net_protocol = REMOTED_NET_PROTOCOL_TCP;
     node_key->rcvd = 10;
     node_key->sock = 15;
+    os_strdup("001",node_key->id);
+
     os_calloc(2, sizeof(keyentry*), keys.keyentries);
     keys.keyentries[0] = node_key;
     keys.keyentries[1] = NULL;
 
     return 0;
 }
 
-static int test_teardown_tcp(void ** state) {
+static int test_teardown_keys(void **state){
     test_mode = 0;
 
+    os_free(keys.keyentries[0]->id);
     os_free(keys.keyentries[0]);
     os_free(keys.keyentries);
 
     return 0;
 }
 
+static int test_setup_tcp(void ** state) {
+
+    test_setup_keys(state);
+    keys.keyentries[0]->net_protocol = REMOTED_NET_PROTOCOL_TCP;
+
+    return 0;
+}
+
+static int test_teardown_tcp(void ** state) {
+    return test_teardown_keys(state);
+}
+
+static int test_setup_udp(void ** state) {
+
+    test_setup_keys(state);
+    keys.keyentries[0]->net_protocol = REMOTED_NET_PROTOCOL_UDP;
+
+    return 0;
+}
+
+static int test_teardown_udp(void ** state) {
+    return test_teardown_keys(state);
+}
+
 /* Tests */
 
+void test_send_msg_invalid_agent(void ** state) {
+    (void) state;
+
+    char *agent_id = "555";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
+
+    remoted_state.queued_msgs = 0;
+
+    expect_function_call(__wrap_pthread_rwlock_rdlock);
+
+    expect_string(__wrap_OS_IsAllowedID, id, agent_id);
+
+    // Setup invalid agent error
+    will_return(__wrap_OS_IsAllowedID, -1);
+
+    expect_function_call(__wrap_pthread_rwlock_unlock);
+
+    expect_string(__wrap__merror, formatted_msg, "(1320): Agent '555' not found.");
+
+    int ret = send_msg(agent_id, msg, msg_length);
+
+    assert_int_equal(ret, -1);
+    assert_int_equal(remoted_state.queued_msgs, 0);
+}
+
+void test_send_msg_disconnected_agent(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
+
+    int key = 0;
+
+    // Setup disconnected agent
+    time_t now = 1000;
+    logr.global.agents_disconnection_time = 300;
+
+    remoted_state.queued_msgs = 0;
+
+    expect_function_call(__wrap_pthread_rwlock_rdlock);
+
+    expect_string(__wrap_OS_IsAllowedID, id, agent_id);
+    will_return(__wrap_OS_IsAllowedID, key);
+    
+    expect_function_call(__wrap_pthread_mutex_lock);
+
+    will_return(__wrap_time, now);
+
+    expect_function_call(__wrap_pthread_mutex_unlock);
+
+    expect_function_call(__wrap_pthread_rwlock_unlock);
+
+    expect_string(__wrap__mdebug1, formatted_msg, "(1245): Sending message to disconnected agent '001'.");
+
+    int ret = send_msg(agent_id, msg, msg_length);
+
+    assert_int_equal(ret, -1);
+    assert_int_equal(remoted_state.queued_msgs, 0);
+}
+
+void test_send_msg_encryption_error(void ** state) {
+    (void) state;
+
+    char *agent_id = "001";
+    char *msg = "abcdefghijk";
+    ssize_t msg_length = 11;
```

Comment:
Hi @jftuduri

Same change here

```suggestion
    ssize_t msg_length = strlen(msg);
```